api/src/main/java/net/wesjd/anvilgui/AnvilGUI.java

```diff
@@ -229,10 +237,21 @@ public void onInventoryClick(InventoryClickEvent event) {
             final int rawSlot = event.getRawSlot();
             if (rawSlot < 3 || event.getAction().equals(InventoryAction.MOVE_TO_OTHER_INVENTORY)) {
                 event.setCancelled(!interactableSlots.contains(rawSlot));
-                final List<ResponseAction> actions =
+                final CompletableFuture<List<ResponseAction>> actionsFuture =
                         clickHandler.apply(rawSlot, StateSnapshot.fromAnvilGUI(AnvilGUI.this));
-                for (final ResponseAction action : actions) {
-                    action.accept(AnvilGUI.this, clicker);
+
+                final Consumer<List<ResponseAction>> actionsConsumer = actions -> {
+                    for (final ResponseAction action : actions) {
+                        action.accept(AnvilGUI.this, clicker);
+                    }
+                };
+
+                if (actionsFuture.isDone()) {
+                    // Fast-path without scheduling if clickHandler is performed in sync
+                    actionsFuture.thenAccept(actionsConsumer);
+                } else {
+                    // Execute the runnable on the main thread
+                    actionsFuture.thenAcceptAsync(actionsConsumer, mainThreadExecutor);
```

Comment:
While adding the ability for async actions is great, theres greater implications we need to consider here:
1. If the inventory is closed by the time the consumer is ran, what do we do?
2. If the plugin is no longer valid to execute the task, what do we do? (Seems like an error for this case)
3. What if the user clicks the completion button multiple times? Should we lock the GUI until the async task finishes? Right now you will spawn as many async tasks as the user clicks the button, which is dangerous.

Reply:
1 -> We should still execute the ResponseActions I'd say, because the plugin developer would expect that they are actually run, even if the AnvilGUI is closed. For example he could have returned an ResponseAction.openInventory() that is no longer executed.
2 -> The Bukkit Scheduler would throw an exception, saying that tasks cannot be scheduled for disabled plugins when using the executor.
We could do an thenAccept() where we first check if the plugin is still enabled in the async thread that handed over execution to AnvilGUI and only then launch a nested .thenAcceptAsync on the main thread executor.
This would be mostly only an issue during a server shutdown, because otherwise a plugin stays enabled.
3 -> The spawning of the async tasks would be performed by the consuming plugin anyway and (for me it would be reasonable) that the plugin should handle that case by itself. Although this requirement should then be documented in the Builder.onClickAsync method. A somewhat dirty workaround that I can think of would be to have an AtomicBoolean stored somewhere that is set to true while any click-handler is running and while true prevents further calls to the handler

Reply:
Nr 2 would look like this:
```suggestion
                    actionsFuture.thenAccept(actions -> {
                        if (plugin.isEnabled()) {
                            // Execute the consumer on the main thread
                            mainThreadExecutor.execute(() -> actionsConsumer.accept(actions));
                        }
                    });
```